### PR TITLE
Fetcher improvements

### PR DESCRIPTION
**Pull request by lheckemann**
###### Motivation for this change
More clarity in store path names, more abstraction

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by lheckemann** [comment]
Eval failure exists on staging as well.

**Review by grahamc — COMMENTED** [review]
Note that the sources are named "source" on purpose, see:

 - https://github.com/nixos/nixpkgs/commit/c3255fe8ec32
 - https://github.com/NixOS/nix/commit/65b5f177b5fbb1b0778ede047a13a3cee9c59cfe
 - https://github.com/NixOS/nix/issues/904

and I think @domenkozar did a talk on why this was a good change. We should talk a bit on if it is a good idea to go back to the previous way. cc @edolstra

**Comment by grahamc** [comment]
One of the points raised in favor of this is the idea that the `name` parameter would make it harder to update the name of a package, but not the source's hash. Unfortunately, that doesn't actually work. The name of the package doesn't have any bearing on the `sha256`:

```
let
  pkgs = import <nixpkgs> {};

  srcWithName = name: pkgs.fetchgit {
    inherit name;
    url = "https://github.com/grahamc/ish.git";
    rev = "7edd7665edf8e33e6847d891d2ff519b492b32dc";
    sha256 = "0qjp8f82a09spz8z2idqj3w1lclr19p37877a8jscvss5a14bbda";
  };

in pkgs.writeText "example" ''
    ${srcWithName "hi"}
    ${srcWithName "there"}
    ${srcWithName "the"}
    ${srcWithName "hash"}
    ${srcWithName "is"}
    ${srcWithName "unchanged"}
  ''
```

**Comment by grahamc** [comment]
@GrahamcOfBorg eval

File: pkgs/build-support/fetchgithub/default.nix [hunk 1]
```diff
@@ -1,6 +1,6 @@
 { lib, fetchgit, fetchzip }:
 
-{ owner, repo, rev, name ? "source"
+{ owner, repo, rev, name ? "${repo}-${builtins.substring 0 10 rev}"
```
Comment (volth) [inline pkgs/build-support/fetchgithub/default.nix]: There was something important in those ```"source"``` defaults but I forgot.
@orivej ?

**Comment by grahamc** [comment]
Note the eval error on staging was fixed, and the error now is due to changes in this PR.

**Comment by edolstra** [comment]
Indeed the use of `source` is by design. We definitely don't want to go back to the situation where every fetcher (and Hydra) produced incompatible store paths.

**Pull request closed without merging**